test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,501 @@
+    end)
+
+    g.server_1:box_config({
+        replication_synchro_timeout = 0.1,
+    })
+
+    g.server_2:box_config({
+        replication_synchro_quorum = 3,
+        replication_synchro_timeout = 1000,
+    })
+
+    local lsn = g.server_2:get_lsn()
+    g.server_2:exec(function()
+        local s = box.space.test
+        require('fiber').create(s.replace, s, {1}):id()
+    end)
+    g.server_1:wait_lsn(g.server_2, lsn + 1)
+
+    -- Start demoting server_2 and interrupt it from server_1
+    local f = demote_start(g.server_2)
+    local election_term = g.server_1:election_term()
+    g.server_1:exec(function()
+        pcall(box.ctl.promote)
+    end)
+    g.server_1:wait_election_term(election_term + 1)
+    --@TODO sometimes this causes assert on server_2
+    -- src/box/txn_limbo.c:515: txn_limbo_read_promote:
+    -- Assertion `txn_limbo_is_empty(&txn_limbo)' failed.
+    -- https://github.com/tarantool/tarantool/issues/6842
+    local ok, err = fiber_join(g.server_2, f)
+
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering demote not handled')
+end
+
+g_spaces.after_test('test_limbo_full_interfering_demote', function(g)
+    g.cluster:promote(g.server_1)
+    g.server_1:exec(function()
+        box.space.test:drop()
+    end)
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Promoting should fail if synchro queue replication timeouts during it
+g_spaces.test_fail_limbo_ack_promote = function(g)
+    g.server_1:box_config({
+        replication_synchro_quorum = 3,
+    })
+
+    g.server_2:box_config({
+        replication_synchro_quorum = 3,
+        replication_synchro_timeout = 1000,
+    })
+
+    -- Fill synchro queue on server_1
+    g.server_2:exec(function()
+        box.ctl.promote()
+        local s = box.schema.create_space('test', {
+            is_sync = true, engine = os.getenv('TARANTOOL_ENGINE')
+        })
+        s:create_index('pk')
+        require('fiber').create(s.replace, s, {1}):id()
+    end)
+    g.server_1:wait_lsn(g.server_2)
+
+    -- Start promoting with default replication_synchro_timeout,
+    -- wait until promote reaches waiting for limbo_acked,
+    -- make it timeout by lowering replication_synchro_timeout
+    local ok, err = g.server_1:exec(function()
+        local fiber = require('fiber')
+        local f = fiber.new(function() box.ctl.promote() end)
+        f:set_joinable(true)
+        box.cfg{replication_synchro_timeout = 0.1}
+        return f:join()
+    end)
+
+    luatest.assert(not ok and err.code == box.error.QUORUM_WAIT,
+        'wait quorum failure not handled')
+end
+
+g_spaces.after_test('test_fail_limbo_ack_promote', function(g)
+    g.server_2:exec(function()
+        -- Timeout for rollback
+        box.cfg{replication_synchro_timeout = 0.1}
+        -- Wait for rollback to happen
+        require('fiber').sleep(0.1)
+        -- Just in case we are not leader somehow
```

Comment:
Is it possible?

Reply:
bad comment here. It is not only possible, it will happen. server_1 will increment term and then fail its promote.